src/clients/patreon/patreonAgent.ts

```diff
@@ -0,0 +1,235 @@
+      patreonUserId: patronDetails.patreonUserId,
+    });
+    if (
+      patreonAccountInUse &&
+      patreonAccountInUse.proavalonUsernameLower !== usernameLower
+    ) {
+      throw new Error(
+        'Attempted to upload a used Patreon for more than one user.',
+      );
+    }
+
+    let result: PatreonDetails;
+
+    if (patronDetails.patreonMemberDetails) {
+      // They are a current member
+      result = await this.updateCurrentPatreonMember(
+        existingPatreon,
+        patronDetails,
+        usernameLower,
+        tokens,
+      );
+    } else {
+      // They are not a current member to the Patreon page
+      result = await this.updateCurrentNonPatreonMember(
+        existingPatreon,
+        tokens,
+        usernameLower,
+        patronDetails.patreonUserId,
+      );
+    }
+
+    console.log(
+      `Successfully linked Patreon account: proavalonUsernameLower="${usernameLower}" patreonUserId="${patronDetails.patreonUserId}" patreonName="${patronDetails.patreonUsersName}" isActivePatreon="${result.isActivePatreon}" amountCents="${result.amountCents}"`,
+    );
+
+    return result;
+  }
+
+  private async updateCurrentPatreonMember(
+    existingPatreon: any,
+    patronDetails: any,
+    usernameLower: string,
+    tokens: PatreonUserTokens,
+  ): Promise<PatreonDetails> {
+    const amountCents =
+      patronDetails.patreonMemberDetails.currently_entitled_amount_cents;
+    const lastChargeDate = new Date(
+      patronDetails.patreonMemberDetails.last_charge_date,
+    );
+
+    // Check payment received
+    const thirtyDaysAgo = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
+    const hasPaid =
+      patronDetails.patreonMemberDetails.last_charge_status &&
+      patronDetails.patreonMemberDetails.last_charge_status === 'Paid' &&
+      lastChargeDate &&
+      lastChargeDate > thirtyDaysAgo;
+    const currentPledgeExpiryDate = hasPaid
+      ? patronDetails.patreonMemberDetails.next_charge_date
+      : null;
+
+    const patreonUpdateDetails = {
+      patreonUserId: patronDetails.patreonUserId,
+      patreonUsersName: patronDetails.patreonUsersName,
+      proavalonUsernameLower: usernameLower,
+      userAccessToken: tokens.userAccessToken,
+      userRefreshToken: tokens.userRefreshToken,
+      userAccessTokenExpiry: tokens.userAccessTokenExpiry,
+      amountCents,
+      // TODO-kev: Check this is accurate
+      currentPledgeExpiryDate: currentPledgeExpiryDate,
+    };
+
+    if (existingPatreon) {
+      await patreonId.findOneAndUpdate(
+        {
+          proavalonUsernameLower: usernameLower,
+        },
+        patreonUpdateDetails,
+      );
+    } else {
+      await patreonId.create(patreonUpdateDetails);
+    }
+
+    return {
+      patreonUserId: patronDetails.patreonUserId,
+      patreonUsersName: patronDetails.patreonUsersName,
+      isActivePatreon: !this.hasExpired(currentPledgeExpiryDate),
+      amountCents,
+    };
+  }
+
+  private async updateCurrentNonPatreonMember(
```

Comment:
We probably don't need this? If the getPatronDetails comes back without patreon member info, then we should just remove their record?